src/source/Ice/SocketConnection.c

```diff
@@ -155,37 +160,75 @@ STATUS socketConnectionSendData(PSocketConnection pSocketConnection, PBYTE pBuf,
 
     MUTEX_LOCK(pSocketConnection->lock);
     locked = TRUE;
 
     if (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP && pSocketConnection->secureConnection) {
-        // underlying BIO has been bound to socket so SSL_write sends the encrypted data.
-        // because we are using non-blocking socket, SSL_write may return SSL_ERROR_WANT_WRITE in which case
-        // SSL_write should be called again with the same buffer.
-        while(iterate) {
+        if (SSL_is_init_finished(pSocketConnection->pSsl)) {
+            /* Should have a valid buffer */
+            CHK(pBuf != NULL && bufLen > 0, STATUS_INVALID_ARG);
             sslRet = SSL_write(pSocketConnection->pSsl, pBuf, bufLen);
-            if (sslRet > 0 || (sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet)) != SSL_ERROR_WANT_WRITE) {
-                iterate = FALSE;
-            } else {
-                THREAD_SLEEP(SSL_WRITE_RETRY_DELAY);
+            if (sslRet < 0){
+                sslErr = SSL_get_error(pSocketConnection->pSsl, sslRet);
+                switch (sslErr) {
+                    case SSL_ERROR_WANT_READ:
+                        /* explicit fall-through */
+                    case SSL_ERROR_WANT_WRITE:
+                        break;
+                    default:
+                        DLOGW("SSL_write failed with %s", ERR_error_string(sslErr, NULL));
+                        DLOGD("Close socket %d", pSocketConnection->localSocket);
+                        ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, TRUE);
+                        break;
+                }
+
+                CHK(FALSE, STATUS_SEND_DATA_FAILED);
             }
         }
-        if (sslRet < 0 && sslErr != SSL_ERROR_WANT_READ) {
-            DLOGW("SSL_write failed with %s", ERR_error_string(sslErr, NULL));
+
+        wBioDataLen = (SIZE_T) BIO_get_mem_data(SSL_get_wbio(pSocketConnection->pSsl), &wBioBuffer);
+        CHK_ERR(wBioDataLen >= 0, STATUS_SEND_DATA_FAILED, "BIO_get_mem_data failed");
+        if (wBioDataLen > 0 && send(pSocketConnection->localSocket, wBioBuffer, wBioDataLen, NO_SIGNAL) < 0) {
+            DLOGE("send data failed with errno %s", strerror(errno));
+
+            if (!socketShouldRetry(errno)) {
+                DLOGD("Close socket %d", pSocketConnection->localSocket);
+                ATOMIC_STORE_BOOL(&pSocketConnection->connectionClosed, TRUE);
+            }
+
+            /* need to also reset BIO on negative path */
+            retStatus = STATUS_SEND_DATA_FAILED;
         }
 
+        /* reset bio to clear its content since it's already sent */
+        BIO_reset(SSL_get_wbio(pSocketConnection->pSsl));
+
     } else if (pSocketConnection->protocol == KVS_SOCKET_PROTOCOL_TCP) {
-        if (send(pSocketConnection->localSocket, pBuf, bufLen, 0) < 0) {
+        /* Should have a valid buffer */
+        CHK(pBuf != NULL && bufLen > 0, STATUS_INVALID_ARG);
```

Comment:
Why is this check not done once in the beginning?

Reply:
because I also use this function to send handshake bits that ssl generates internally after SSL_do_handshake. In that case there is no application data.

Reply:
So, in that case we still have the size param but buffer is NULL?

Reply:
size is 0 and buffer is NULL.

Reply:
I see.. Do we need to add a check for when size is not 0 then buffer should not be NULL?

Reply:
so it's either SSL init is not finished in which case pBuf and bufLen are completely ignored, or it's sending application data and thus pBuf != NULL and bufLen > 0.